homeassistant/components/generic_thermostat/climate.py

```diff
@@ -43,84 +45,190 @@
+                vol.Optional(CONF_ECO_SHIFT): vol.Coerce(float),
+                vol.Optional(CONF_COMFORT_SHIFT): vol.Coerce(float),
+            }
+        ),
+        vol.Optional(HVAC_MODE_COOL): vol.Schema(
+            {
+                vol.Required(CONF_ENTITY_ID): cv.entity_id,
+                vol.Required(
+                    CONF_HVAC_MODE_MIN_TEMP, default=DEFAULT_MIN_TEMP_COOL
+                ): vol.Coerce(float),
+                vol.Required(
+                    CONF_HVAC_MODE_MAX_TEMP, default=DEFAULT_MAX_TEMP_COOL
+                ): vol.Coerce(float),
+                vol.Required(
+                    CONF_HVAC_MODE_INIT_TEMP, default=DEFAULT_TARGET_TEMP_COOL
+                ): vol.Coerce(float),
+                vol.Optional(CONF_AWAY_TEMP): vol.Coerce(float),
+                vol.Optional(CONF_ECO_SHIFT): vol.Coerce(float),
+                vol.Optional(CONF_COMFORT_SHIFT): vol.Coerce(float),
+            }
         ),
-        vol.Optional(CONF_AWAY_TEMP): vol.Coerce(float),
-        vol.Optional(CONF_PRECISION): vol.In(
-            [PRECISION_TENTHS, PRECISION_HALVES, PRECISION_WHOLE]
+        vol.Optional(CONF_ENABLED_PRESETS, default=[]): vol.All(
+            cv.ensure_list, SUPPORTED_PRESET_MODES
         ),
+        vol.Optional(CONF_MIN_CYCLE_DURATION): vol.All(
+            cv.time_period, cv.positive_timedelta
+        ),
+        vol.Optional(CONF_ENABLE_OLD_STATE, default=DEFAULT_OLD_STATE): cv.boolean,
     }
 )
 
 
 async def async_setup_platform(hass, config, async_add_entities, discovery_info=None):
     """Set up the generic thermostat platform."""
     name = config.get(CONF_NAME)
-    heater_entity_id = config.get(CONF_HEATER)
     sensor_entity_id = config.get(CONF_SENSOR)
-    min_temp = config.get(CONF_MIN_TEMP)
-    max_temp = config.get(CONF_MAX_TEMP)
-    target_temp = config.get(CONF_TARGET_TEMP)
-    ac_mode = config.get(CONF_AC_MODE)
-    min_cycle_duration = config.get(CONF_MIN_DUR)
-    cold_tolerance = config.get(CONF_COLD_TOLERANCE)
-    hot_tolerance = config.get(CONF_HOT_TOLERANCE)
-    keep_alive = config.get(CONF_KEEP_ALIVE)
     initial_hvac_mode = config.get(CONF_INITIAL_HVAC_MODE)
-    away_temp = config.get(CONF_AWAY_TEMP)
-    precision = config.get(CONF_PRECISION)
     unit = hass.config.units.temperature_unit
+    initial_preset_mode = config.get(CONF_INITIAL_PRESET_MODE)
+    tolerance_on = config.get(CONF_HYSTERESIS_TOLERANCE_ON)
+    tolerance_off = config.get(CONF_HYSTERESIS_TOLERANCE_OFF)
+    keep_alive = config.get(CONF_KEEP_ALIVE)
+    enabled_presets = config.get(CONF_ENABLED_PRESETS)
+    min_cycle_duration = config.get(CONF_MIN_CYCLE_DURATION)
+    enable_old_state = config.get(CONF_ENABLE_OLD_STATE)
+    heat_conf = config.get(HVAC_MODE_HEAT)
+    cool_conf = config.get(HVAC_MODE_COOL)
+    enabled_hvac_modes = []
+
+    # Do all the checks I cannot do with voluptuous
+    if initial_preset_mode != "none" and initial_preset_mode not in enabled_presets:
```

Comment:
You should follow homeassistant.helpers.config_validation as an example to do this check

Reply:
Thanks for this information, is this the code located in this file ? https://github.com/home-assistant/home-assistant/blob/master/homeassistant/helpers/config_validation.py

Because I've read it now, and I still don't know what I'm supposed to do. Can you point me in the right direction ?

Reply:
I didn't quite get your comment on L160 

Can't you extent the config_validation with new functions to do these checks through voluptuous ?

Reply:
Unfortunately I have no idea how I can do such thing. If you know a component that do something like that, please share its name with me so I can have an example of what I'm supposed to do.

Reply:
Just picking up a reference to the first extension in config_validation.py (https://github.com/home-assistant/home-assistant/blob/75b508e8724cb82712227082eee9abab570d2eb2/homeassistant/helpers/config_validation.py#L94)

In azure_service_bus component this extra validation is used: 

https://github.com/home-assistant/home-assistant/blob/75b508e8724cb82712227082eee9abab570d2eb2/homeassistant/components/azure_service_bus/notify.py#L32

Reply:
Thanks, I did an attempt to implement the first check in voluptuous as a custom validator. It's working. Tell me if it is what you are expecting, and I will do it for the next checks.

Reply:
I did all the checks with voluptuous